### ISSUE TITLE
Add AD version of the Moose Function system

### ISSUE THREAD

**Issue by dschwen** [idaholab/moose#14262]
## Reason
When evaluating moose function objects with dual number arguments derivatives have to be dropped. This kills a lot of applications for functions when using the automatic differentiation capabilities. Multilinear functions could be used as lookup tables in state space (indexed by non-linear variable values).

## Design
Have @lindsayad develop an awesome design. E.g. new ADFunction base class and new methods (and macros) in FunctionInterface.

## Impact
Added capability.

**Comment by lindsayad** [comment]
Hahaha I love the design description

**Comment by lindsayad** [comment]
> When evaluating moose function objects with dual number arguments derivatives have to be dropped. This kills a lot of applications for functions when using the automatic differentiation capabilities. Multilinear functions could be used as lookup tables in state space (indexed by non-linear variable values).

@dschwen I guess I don't fully understand what's wanted here. The `Function` APIs just takes the time and a physical point in space. What are you referring to with the `DualNumber` arguments?

**Comment by dschwen** [comment]
Functions are frequently abused by passing nonlinear variables in for time.

**Comment by lindsayad** [comment]
So `ADFunction` should have an API different from that of our regular `Functions`? This seems to me like we are getting more into `ParsedFunction` type territory where we might want to accept variadic arguments.

BTW @dschwen can you respond to [this](https://groups.google.com/forum/?nomobile=true#!topic/moose-users/UtXfmsN4cqU)?

**Comment by lindsayad** [comment]
I feel like this issue should be lumped with #13109, perhaps under a more general title. @dschwen your thought?

**Comment by dschwen** [comment]
> I feel like this issue should be lumped with #13109

no

**Comment by lindsayad** [comment]
Ok, with the limited info I have here and without creeping into `ParsedFunction`, my API idea for this awesome new base class is:
```
ADReal value(const ADReal & time_like_arg, const VectorValue<ADReal> & point_like_arg);
```
Is that all you want? With corresponding `gradient, etc` APIs?

**Comment by dschwen** [comment]
Ah, I see. I saw this as orthogonal to dual number parsed expressions. The low hanging fruit would be linear interpolation functions. Either way, just getting the interface going with `getADFunction` and `getGenericFunction<is_ad>` and all teh backend work would be a great first step. 

In the meantime my work on #15207 will make it easy to implement parsed AD functions!

**Comment by tophmatthews** [comment]
Maybe think about using splines or smootherstep rather than linear interpolation? It may be hard to get good convergence with a discontinuous first derivative, no?

**Comment by dschwen** [comment]
It was just an example of a function that could be easily converted.
Ultimately we should have as and non as versions of every function.

**Comment by lindsayad** [comment]
You want automatic scaling (as) versions? :-)

**Comment by lindsayad** [comment]
@dschwen How about:
```c++
template <std::size_t N1, std::size_t N2>
class ADFunction
{
public:
  ADReal value (const std::array<const ADReal *, N1> & scalars,
                          const std::array<const VectorValue<ADReal> *, N2> & vectors);
};
```
or alternatively just:
```c++
template <std::size_t N>
class ADFunction
{
public:
  ADReal value (const std::array<const ADReal *, N> & args);
};
```
?

**Comment by dschwen** [comment]
Closed by #20200
